File: backend/app/schemas/capacity.py
"""Schemas — análise de capacidade e Erlang C (R-C)."""

from pydantic import BaseModel, Field


class ResourceSection(BaseModel):
    cpu_cores: float
    cpu_percent_used: float
    cpu_available_ratio: float
    ram_total_mb: float
    ram_available_mb: float
    gpu_signal_available: bool
    gpu_signal_source: str | None = None
    gpu_device_name: str | None = None
    container_estimate: bool = True


class CapacityEstimateSection(BaseModel):
    resource_units_budget: float
    max_weighted_capacity_estimated: int
    max_weighted_capacity_effective: int
    max_weighted_capacity_override: int = 0
    channels_if_single_family: dict[str, int] = Field(default_factory=dict)
    channel_costs: dict[str, float] = Field(default_factory=dict)
    channel_weights: dict[str, int] = Field(default_factory=dict)
    notes: list[str] = Field(default_factory=list)


class CapacityUsageSection(BaseModel):
    global_usage: int
    global_max: int
    global_remaining: int
    outbound_weight_bound: int
    receptive_weight_bound: int
    unmapped_usage: int


class ObservedTrafficSection(BaseModel):
    period_days: int
    arrival_rate_per_hour: float
    arrival_count: int
    aht_seconds: float
    aht_sample_count: int
    aht_source: str
    traffic_intensity_erlangs: float


class ErlangSection(BaseModel):
    num_agents: int
    traffic_intensity_erlangs: float
    probability_wait: float
    service_level_predicted: float
    service_level_target: float
    service_level_target_seconds: int
    required_agents_for_target: int
    service_level_at_required: float
    headroom_agents: int
    headroom_volume_percent: float
    analytical_only: bool = True


class CapacityResponse(BaseModel):
    """Capacidade estimada + uso global + dimensionamento Erlang C."""

    resources: ResourceSection
    estimate: CapacityEstimateSection
    usage: CapacityUsageSection
    observed: ObservedTrafficSection
    erlang: ErlangSection
    messaging_channels: list[str] = Field(default_factory=list)

File: frontend/src/lib/types/capacity.ts
export interface ResourceSection {
  cpu_cores: number;
  cpu_percent_used: number;
  cpu_available_ratio: number;
  ram_total_mb: number;
  ram_available_mb: number;
  gpu_signal_available: boolean;
  gpu_signal_source: string | null;
  gpu_device_name: string | null;
  container_estimate: boolean;
}

export interface CapacityEstimateSection {
  resource_units_budget: number;
  max_weighted_capacity_estimated: number;
  max_weighted_capacity_effective: number;
  max_weighted_capacity_override: number;
  channels_if_single_family: Record<string, number>;
  channel_costs: Record<string, number>;
  channel_weights: Record<string, number>;
  notes: string[];
}

export interface CapacityUsageSection {
  global_usage: number;
  global_max: number;
  global_remaining: number;
  outbound_weight_bound: number;
  receptive_weight_bound: number;
  unmapped_usage: number;
}

export interface ObservedTrafficSection {
  period_days: number;
  arrival_rate_per_hour: number;
  arrival_count: number;
  aht_seconds: number;
  aht_sample_count: number;
  aht_source: string;
  traffic_intensity_erlangs: number;
}

export interface ErlangSection {
  num_agents: number;
  traffic_intensity_erlangs: number;
  probability_wait: number;
  service_level_predicted: number;
  service_level_target: number;
  service_level_target_seconds: number;
  required_agents_for_target: number;
  service_level_at_required: number;
  headroom_agents: number;
  headroom_volume_percent: number;
  analytical_only: boolean;
}

export interface CapacityResponse {
  resources: ResourceSection;
  estimate: CapacityEstimateSection;
  usage: CapacityUsageSection;
  observed: ObservedTrafficSection;
  erlang: ErlangSection;
  messaging_channels: string[];
}
